server/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -263,13 +280,20 @@ public Throwable getRootCause() {
         return rootCause;
     }
 
+    /**
+     * Write any exception-specific data to {@code out}
+     */
+    protected void writeExceptionDataTo(StreamOutput out) throws IOException {}
```

Comment:
Why is this new method needed? It looks like all the uses of it just replace the existing super + local data calls with using it. Why does writeTo need to be final now?

Reply:
It's a separate method so it can be called on its own to just write the type-specific data, not as part of the `writeTo` method (see call at bottom of the big `writeException` method). I've made `writeTo` final so subclasses don't override it, thinking they can then modify the behaviour every time the exception is serialized - as the logic in the base method is also in `writeException`, which they can't override.

Reply:
What I don't like about this is the method implementation must match that in writeException, but nothing enforces that. I would rather delegate completely to ElasticsearchException in writeException. How about this:
* move this entire implementation to a separate protected method, which takes the stack level, so that subclasses can still override the behavior and
* make this final writeTo method call that protected method with a stack level of 0
* in writeException, after the if/else on types are handled, delegate to that write method if its an ElasticsearchException

Reply:
It also needs to pass through the `Runnable` for the nested stack callback for tests though. That really is an internal implementation detail.

Reply:
I've pushed a change to smuggle the info inside a closure